src/asset-services.ts

```diff
@@ -0,0 +1,123 @@
+function ensureAbsoluteURLs(baseURL: string, entrypoints: string[]) {
+  const regExpIsAbsoluteURL = new RegExp("^(?:[a-z]+:)?//", "i");
+  return entrypoints.map(function (entrypoint) {
+    if (!regExpIsAbsoluteURL.test(entrypoint)) {
+      return baseURL + entrypoint;
+    }
+    return entrypoint;
+  });
+}
+
+function addRef(document: Document, entrypoint: string) {
+  const pattern = /\.([0-9a-z]+)(?=[?#])|(\.)(?:[\w]+)$/gim;
+  switch (entrypoint.match(pattern)![0]) {
```

Comment:
Me explicas como funciona `![0]` al final de esta línea?

Reply:
Sure!

> A new `!` post-fix expression operator may be used to assert that its operand is non-null and non-undefined in contexts where the type checker is unable to conclude that fact. Specifically, the operation `x!` produces a value of the type of `x` with `null` and `undefined` excluded. Similar to type assertions of the forms `<T>x` and `x as T`, the `!` non-null assertion operator is simply removed in the emitted JavaScript code.

> from [Non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator)

Reply:
En todo caso ese es un problema de TypeScript jeje.

Como yo se que nunca `entrypoint.match(pattern)` va a resultar en `null`, con el `!` informo eso al compilador.

Sería similar a esto:
```suggestion
  switch ((entrypoint.match(pattern) as RegExpMatchArray)[0]) {
```